README.rst

```diff
@@ -22,46 +22,96 @@ A pure Python micro-framework supporting seamless lazy and concurrent evaluation
+>>> import paragraph as pg
+>>> import operator
+>>> x, y = pg.Variable("x"), pg.Variable("y")
+>>> add = pg.op(operator.add())
+>>> s = add.op(x, y)
 
-A glossary is provided below, which should clarify most concepts implemented in this module. Note that the usage of some terms may slightly differ from
-their standard definitions, reading it through before diving into the code is therefore highly recommended irrespective of the reader's familiarity with the
-topic.
 
+The few lines above fully instantiate a computation graph, here in its simplest form with just one equation relating ``x``, ``y`` and ``s`` via the function
+``add``. Given values for the input variables ``x`` and ``y``, the value of ``s`` is resolved as follows:
 
-Getting started
-===============
+>>> pg.evaluate([s], {x: 5, y: 10})
+[15]
 
-Computation graphs are `bipartite graphs <https://en.wikipedia.org/wiki/Bipartite_graph>`_, where vertices of a one type, the variables (of type
-Variable), are connected together exclusively through vertices of another type, the operations (or simply ops, of type Op).
 
-In *paragraph*, turning a regular Python function into an op is as simple as decorating it:
+Key features
+''''''''''''
 
->>> @op
-... def f(a, b):
-...     return a * b
+The main benefits of using paragraph stem from the following features of ``pg.session.evaluate``:
 
-The operation can then be applied to objects of both Variable and non-Variable type as follows:
+Lazy evaluation
+  Irrespective of the size of the computation graph, only the operations required to evaluate the output variables are executed. Consider the following
+  extension of the above graph:
 
->>> v1 = Variable("v1")
->>> v2 = f(2, v1)
+  >>> z = pg.Variable("z")
+  >>> t = add.op(y, z)
 
-Ops differ from regular Python functions in their behavior upon receiving an argument of type Variable. In such a case, they are not executed,
-but instead pack the knowledge required for deferred execution into an instance of Variable, and return this variable.
-Then, a variable can be evaluated by invoking the function `paragraph.session.evaluate` and passing in initialization values for the input
-variables alongside the target variable:
+  Then the statement:
 
->>> value = evaluate([v2], args={v1: 4})
+  >>> pg.evaluate([t], {y: 10, z: 50})
+  [60]
 
+  just ignores the variables ``s`` and ``x`` altogether, since they do not contribute to the evaluation of ``t``. In particular, the operation ``add(x, y)``
+  is not executed.
 
-Going further
-=============
+
+Eager currying
+  Invoking an op with invariable arguments (that is, arguments that are not of type ``Variable``) just returns an invariable value: evaluation is
+  eager whenever possible. If only a subset of inputs is a variable, say ``x_p``, the computation graph can be simplified using ``solve``, which returns a new
+  variable:
+  
+  >>> x_p = Variable("x_p")
+  >>> u = add.op(s, t)
+  >>> u_xp = pg.solve([u], {x: x_p, y: 10, z: 50})[0]
```

Comment:
Consider renaming `x_p -> v` and `u_xp -> u_v`. Solving the equation yields `x` as a function of `v`, `x_v`!

Reply:
No, it is yielding ``u`` as a function of ``x_p`` (``_p`` is for "prime").

Reply:
Yes sorry, I mean `u` as a function of `v`, otherwise you'd end up with a double subscript u_{x_{p}} which is not readable.

Reply:
But I should improve that, true. There are two separate messages to be conveyed here: 

1. If some input variables are given a value, but not all, ``solve`` returns a simplified computation graph. For this, ``x_p`` is just useless.
2. If you pass another variable (which may itself depend on anything) as the argument value for the input variable ``x``, ``solve`` returns a new computation graph with the resulting dependencies pieces together. Let me try to make this clearer.

Reply:
Sorry `u_v = solve(u, {x: v, y: 10, z: 50})`  is more readable than the list version `u_v = solve([u], {...})[0]`. At some point the total number of bracket tolerance of the average user gets exhausted...

Reply:
So, I simplified the above example, simply removing the definition of x_p and the argument for x in solve.
I'll add a subsection on "Graph composition" to explain point 2 as a separate functionality (which it is).

Reply:
As for the list thingy, it sure is more readable without the additional brackets. But this is not the only use case, and it is paramount to be able to pass in a list (and I'd expect that to be the most frequent use case, by far).